# Файлы. Частотный словарь

Мы уже знаем про:
- типы данных `int`, `float`, `str`, `bool`, `list` [список], `tuple` (кортеж неизменяемый), `set` {множество}, `dict` словарь {ключ: значение}
- математические операторы: `+`, `-`, `/`, `*`, `**` (в степень), `%` (остаток от деления), `//` (целая часть при делении)
- `print()` — напечатать
- `type()` — узнать тип данных
- `input()` — принять значение от пользователя
- `int()` — преобразовать в целочисленный формат
- `float()` — преобразовать в формат дробного числа
- `str()` — преобразовать в строку
- `bool()` — преобразовать в bool (True/False)
- `list()` — преобразовать в список
- `tuple()` — преобразовать в кортеж
- `set()` — преобразовать во множество
- `dict()` — преобразовать в словарь

- условия:
```
if <условие>:
    <инструкции, для случая если условие истинно (True)>
#elif <условие №2>:
#    <инструкции, для случая если условие №2 истинно (True)>
#else:
#    <инструкции, когда все условия не выполнились (False)>
```
- операторы сравнения: `>`, `<`, `>=`, `<=`, `==` (равно), `!=` (не равно)
- логические операторы: `and`, `or`, `not`
- цикл `while`:
```
while <условие>:
    <инструкции, которые будут выполняться по кругу, пока условие выполняется True>
```
- цикл `for`
```
for <element> in <elements>:
       <инструкции, которые будут выполняться для каждого элемента>
```
- `.split()` - поделить строку, указанную до точки, по сочетанию символов, указанных в скобках (строка -> список)
- индексация с нуля: `итерируемый_объект[индекс]`
- срезы: `итерируемый_объект[START:END:STEP]`
- `*список` - то же самое, что `список[0] список[1] список[2] список[3]...`
- `len()` - узнать длину
- `range()` - "обещание последовательности чисел" (`list(range(0, 8, 2))` - `[0, 2, 4, 6]`)
- `enumerate()` - делает последовательность кортежей: [(порядковый_номер, элемент), (порядковый_номер, элемент), (порядковый_номер, элемент)]

**Методы списка (+ строк и кортежей, если метод ничего не изменяет)**
- `list.append()` - добавление элемента в конец списка
- `list.extend(another_list)` - расширяет список, добавляя в конец все элементы списка `another_list` (и не только списка)
- `list.insert(i, x)` — вставляет в `i`-ую позицию в списке значение `x`
- `list.remove(x)` — удаляет первый элемент в списке, имеющий значение `x`
- `list.pop(i)` - удаляет элемент с индексом `i` (по умолчанию последний) и возвращает удалённый элемент
- `list.clear()` — очищает список
- `list.count(x)` — узнать, сколько раз встречается элемент `x`
- `list.index(x)` — узнать индекс элемента `x` в списке

**Функции для списков и кортежей**
- `sum()` — сумма
- `min()` — минимальное
- `max()` — максимальное
- `map()` — применить функцию или метод ко всем элементам сразу (`print(list(map(len, ['Привет', 'ТЕСТ', 'функции', 'map!'])))` - `[6, 4, 7, 4]`), (`print(list(map(str.split, ['Таня Казакова', 'Максим Бажуков', 'Олег Сериков'])))` - `[['Таня', 'Казакова'], ['Максим', 'Бажуков']]`)

**Методы строк**
- `str.strip()`, `str.lstrip()`, `str.rstrip()` - убрать символы на концах строки
- `str.lower()`, `str.casefold()` - понизить регистр
- `str.capitalize()`, `str.upper()`, `str.title()` - повысить регистр (как в предложении, всех букв, начальной буквы каждого слова)
- `str.replace("<что>", "<на что>")` - делаем замены
- `str.find()` (`str.rfind()`) (а также `str.index()`, `str.rindex()`) - найти позицию символа в строке
- `str.count()` - подсчёт нужных символов в строке
- `str.split()` с `maxsplit` - делит строку по выбранной строке
- `<чем>.join(<что>)` - соединить элементы списка
- `str.startswith()` - проверяет, что строка начинается с определённой подстроки
- `str.endswith()` - проверяет, что строка кончается на определённую подстроку
- `str.isalpha()` - проверяет, что строка состоит только из букв
- `str.isdigit()` - проверяет, что строка состоит только из цифр
- `str.isalnum()` - проверяет, что строка состоит только из букв и цифр
- `str.islower()` - проверяет, что все буквы в строке нижнего регистра
- `str.isupper()` - проверяет, что все буквы в строке верхнего регистра

Форматирование строк 1:`"Результат: {}".format(res)`

Форматирование строк 2: `f"Результат: {res}"`

**Методы множеств**
- `.add()` - добавить элемент
- `.update()` - добавить несколько элементов в множество
- `.remove()` - удалить элемент из множества
- `.union()` или `|` - объединяем множества
- `.intersection()` или `&` - пересекаем множества
- `.difference()` или `-` - ищем разность
- `.symmetric_difference()` или `^` - находим симметрическую разность

**Словари**
- `eng_rus_dict['house']` - узнать значение по ключу
- `.keys()` — возвращает объект с ключами словаря.
- `.values()` — возвращает объект со значениями словаря.
- `.items()` — возвращает объект с кортежами вида (ключ, значение)
- `.get(<ключ>, <что вывести, если ключ не найден>)` — возвращает значение по ключу или то, что надо вывести, если ключ не найден
- `eng_rus_dict["new"] = 'новый'` - добавить новую пару ключ-значение или изменить существующее значение у ключа
- `del eng_rus_dict['key']` - удалить пару ключ-значение

**Функции и методы для коллекций**
- `.copy()` - сделать копию коллекции данных
- `sorted(<что сортировать>, key=<какую функцию применить к аргументам>)` — функция сортировки (например: sorted(words, key=str.lower))

**Изменяемость типов данных**

Списки, множества и словари - изменяемые.

**Общий вид объявления функции**
```
def <название_функции>(): 
    <тело/инструкции_функции>
    return <результат работы функции>
```
**Модули и их функции**
- `import math` - импортируем модуль math
- `from math import sqrt` - импортируем функцию модуль `sqrt` из `math`
- `from math import *` - импортируем всё из модуля `math`
- `import string as s` - даем псевдоним модулю `string`
- `math.sqrt()` - вычисляет квадратный корень
- `random.randint()` - даёт рандомное целое число из промежутка
- `random.uniform()` - даёт рандомное дробное число из промежутка
- `string.punctuation` - частотные пунктуационные знаки

**Файлы**

Общий вид:
```
with open('first_text.txt', mode='r', encoding='UTF-8') as f:
    text = f.read()
```
Режимы (mode): `r` - чтение, `w` - перезапись, `a` - дозапись 

Кодировка: `encoding='UTF-8'` (иногда бывает `windows-1251` и др.)

## Большая тренировка
Прочитайте "Войну и мир", выдайте 20 самых частотных слов
- сохранить текст в переменную
- разбить на слова, привести каждое слово к нижнему регистру, очистить от пунктуации
- убрать стоп-слова
- сделать словарь слово-количество употреблений
- отсортировать слова по количеству
- вывести 20 частотных слов

In [1]:
with open('data/vojna-i-mir.txt', 'r', encoding='UTF-8') as f:
    text = f.read()
print(text[:50])



Том первый


Часть первая



I

— Eh bien, mon p


In [2]:
from string import punctuation

In [3]:
punctuation + '—'

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~—'

In [4]:
words = []
for word in text.split():
    good_word = word.strip(punctuation + '—').lower()
    if good_word != '':
        words.append(good_word)
words[:10]

['том',
 'первый',
 'часть',
 'первая',
 'i',
 'eh',
 'bien',
 'mon',
 'prince',
 'gênes']

### Частотный словарь
Хотим словарь `{'слово1': <количество>, 'слово2': <количество>...}`

Как делать не надо?

In [5]:
# Очень долго
uniquew = set(words)
dict_w_num = {}
for word in uniquew:
    dict_w_num[word] = words.count(word)

In [6]:
len(uniquew)

57994

In [7]:
len(words)

484531

Мы попытались пройти по множеству из 57994 слов и для каждого пройтись по списку всех слов (их 484531), чтобы каждое отдельно посчитать. Это безумно долго.

Гораздо быстрее будет пройтись по списку слов только один раз!

In [8]:
# Если слово ещё не встречалось, запиши, что оно встретилось впервые.
# А если уже встречалось, запиши, что мы ещё один раз его встретили.

word2number = {}
for word in words:
    if word not in word2number:
        word2number[word] = 1
    else:
        word2number[word] += 1

In [9]:
word2number['наташа']

821

Но как вывести 20 пар из словаря?

In [10]:
list(word2number.items())[:20]

[('том', 789),
 ('первый', 182),
 ('часть', 87),
 ('первая', 37),
 ('i', 76),
 ('eh', 29),
 ('bien', 64),
 ('mon', 187),
 ('prince', 37),
 ('gênes', 2),
 ('et', 306),
 ('lucques', 2),
 ('ne', 146),
 ('sont', 33),
 ('plus', 72),
 ('que', 277),
 ('des', 113),
 ('apanages', 1),
 ('поместья', 4),
 ('de', 626)]

Как теперь слова отсортировать по количеству?

`sorted()` умеет сортировать такой список по нулевому элементу в кортежах. Тогда надо поменять местами, чтоб сортировать не по словам, а по числу употреблений.

In [11]:
num_word = []
for tupl in list(word2number.items()):
    num_word.append((tupl[1], tupl[0]))

In [12]:
sorted(num_word, reverse=True)[:20]

[(22036, 'и'),
 (11802, 'в'),
 (9054, 'не'),
 (7988, 'что'),
 (7622, 'он'),
 (6985, 'на'),
 (6171, 'с'),
 (4209, 'как'),
 (4058, 'его'),
 (3599, 'к'),
 (3268, 'я'),
 (2857, 'но'),
 (2769, 'она'),
 (2693, 'все'),
 (2686, 'это'),
 (2565, 'было'),
 (2373, 'то'),
 (2063, 'а'),
 (2027, 'сказал'),
 (2026, 'так')]

Вообще-то мы могли бы сделать то же самое меньшими усилиями.
    Сделать с помощью `.items()` список кортежей и сортировать его, по специальному ключу `key`!
    Это должен быть ключ, который вернёт из кортежа второй элемент &mdash; частоту слова.
    Так и напишем:

In [13]:
def get_second_element(indexable_collection):
    return indexable_collection[1]

In [14]:
sorted(word2number.items(),
       key=get_second_element,
       reverse=True
      )[:20]

[('и', 22036),
 ('в', 11802),
 ('не', 9054),
 ('что', 7988),
 ('он', 7622),
 ('на', 6985),
 ('с', 6171),
 ('как', 4209),
 ('его', 4058),
 ('к', 3599),
 ('я', 3268),
 ('но', 2857),
 ('она', 2769),
 ('все', 2693),
 ('это', 2686),
 ('было', 2565),
 ('то', 2373),
 ('а', 2063),
 ('сказал', 2027),
 ('так', 2026)]

Это самые частотные слова. Но они во всех длинных текстх частотны. Как бы оставить только содержательное про текст?

### Стоп-слова

Стоп-слова - слова самые частотные во всех текстах на определённом языке.

Обычно это союзы, предлоги, частицы, местоимения, артикли (не в русском).

Их можно убрать убрать.

In [15]:
! pip install nltk

DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621

[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python3.9 -m pip install --upgrade pip


In [16]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/tbkazakova/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [17]:
from nltk.corpus import stopwords
print(stopwords.words('russian'))

['и', 'в', 'во', 'не', 'что', 'он', 'на', 'я', 'с', 'со', 'как', 'а', 'то', 'все', 'она', 'так', 'его', 'но', 'да', 'ты', 'к', 'у', 'же', 'вы', 'за', 'бы', 'по', 'только', 'ее', 'мне', 'было', 'вот', 'от', 'меня', 'еще', 'нет', 'о', 'из', 'ему', 'теперь', 'когда', 'даже', 'ну', 'вдруг', 'ли', 'если', 'уже', 'или', 'ни', 'быть', 'был', 'него', 'до', 'вас', 'нибудь', 'опять', 'уж', 'вам', 'ведь', 'там', 'потом', 'себя', 'ничего', 'ей', 'может', 'они', 'тут', 'где', 'есть', 'надо', 'ней', 'для', 'мы', 'тебя', 'их', 'чем', 'была', 'сам', 'чтоб', 'без', 'будто', 'чего', 'раз', 'тоже', 'себе', 'под', 'будет', 'ж', 'тогда', 'кто', 'этот', 'того', 'потому', 'этого', 'какой', 'совсем', 'ним', 'здесь', 'этом', 'один', 'почти', 'мой', 'тем', 'чтобы', 'нее', 'сейчас', 'были', 'куда', 'зачем', 'всех', 'никогда', 'можно', 'при', 'наконец', 'два', 'об', 'другой', 'хоть', 'после', 'над', 'больше', 'тот', 'через', 'эти', 'нас', 'про', 'всего', 'них', 'какая', 'много', 'разве', 'три', 'эту', 'моя', 'впр

In [18]:
important_words = []
for tupl in num_word:
    if tupl[1] not in stopwords.words('russian'):
        important_words.append(tupl)
sorted(important_words, reverse=True)[:20]

[(2686, 'это'),
 (2027, 'сказал'),
 (1413, 'пьер'),
 (1385, 'князь'),
 (821, 'наташа'),
 (809, 'андрей'),
 (748, 'время'),
 (715, 'сказала'),
 (709, 'которые'),
 (674, 'который'),
 (671, 'говорил'),
 (626, 'de'),
 (611, 'очень'),
 (576, 'княжна'),
 (549, 'своей'),
 (530, 'ростов'),
 (524, 'человек'),
 (513, 'марья'),
 (501, 'лицо'),
 (483, 'граф')]

Ну вот, уже содержательнее.

(Есть более умные способы искать ключевые слова. Может, сможем посмотреть их чуть позже.)